src/CBLDatabase.cc

```diff
@@ -70,6 +71,165 @@ bool CBLEncryptionKey_FromPasswordOld(CBLEncryptionKey *key, FLString password)
+// CBLCollection_Internal.hh, which would create a circular header dependency.
+
+
+#pragma mark - CONSTRUCTORS:
+
+
+CBLDatabase::CBLDatabase(C4Database* _cbl_nonnull db, slice name_, slice dir_)
+:_c4db(std::move(db))
+,_dir(dir_)
+,_notificationQueue(this)
+{ }
+
+
+CBLDatabase::~CBLDatabase() {
+    _c4db.useLocked([&](Retained<C4Database> &c4db) {
+        // Invalidate the database reference for both scopes and collections:
+        for (auto& i : _scopes) {
+            i.second->close();
+        }
+        
+        for (auto& i : _collections) {
+            i.second->close();
+        }
+    });
+}
+
+
+#pragma mark - SCOPES:
+
+
+CBLScope* CBLDatabase::getScope(slice scopeName) {
+    if (!scopeName)
+        scopeName = kC4DefaultScopeID;
+    
+    auto c4db = _c4db.useLocked();
+    
+    // TODO: change to use hasScope()
+    bool exist = (scopeName == kC4DefaultScopeID); // Default scope always exist.
+    if (!exist) {
+        c4db->forEachScope([&](slice scope) {
+            if (!exist && scopeName == scope)
+                exist = true;
+        });
+    }
+    
+    CBLScope* scope = nullptr;
+    if (auto i = _scopes.find(scopeName); i != _scopes.end()) {
+        if (exist) {
+            scope = i->second.get();
+        } else {
+            _scopes.erase(i);
+        }
+    }
+    
+    if (exist && !scope) {
+        auto retainedScope = make_retained<CBLScope>(scopeName, this);
+        scope = retainedScope.get();
+        _scopes.insert({scope->name(), move(retainedScope)});
+    }
+    return scope;
+}
+
+
+#pragma mark - COLLECTIONS:
+
+
+CBLCollection* CBLDatabase::getCollection(slice collectionName, slice scopeName) const {
+    if (!scopeName)
+        scopeName = kC4DefaultScopeID;
+
+    auto c4db = _c4db.useLocked();
+    auto spec = C4Database::CollectionSpec(collectionName, scopeName);
+    auto c4col = c4db->getCollection(spec);
+    if (!c4col) {
+        removeCBLCollection(spec);
+        return nullptr;
+    }
+    return getOrCreateCBLCollection(c4col);
+}
+
+
+CBLCollection* CBLDatabase::createCollection(slice collectionName, slice scopeName) {
+    if (!scopeName)
+        scopeName = kC4DefaultScopeID;
+    
+    auto c4db = _c4db.useLocked();
+    auto spec = C4Database::CollectionSpec(collectionName, scopeName);
+    auto c4col = c4db->createCollection(spec);
+    return getOrCreateCBLCollection(c4col);
+}
+
+
+bool CBLDatabase::deleteCollection(slice collectionName, slice scopeName) {
+    if (!scopeName)
+        scopeName = kC4DefaultScopeID;
+    
+    auto c4db = _c4db.useLocked();
+    auto spec = C4Database::CollectionSpec(collectionName, scopeName);
+    c4db->deleteCollection(spec);
+    removeCBLCollection(spec);
+    return true;
+}
+
+
+CBLCollection* CBLDatabase::getDefaultCollection() const {
+    return getDefaultCollection(false);
+}
+
+
+CBLCollection* CBLDatabase::getDatabaseDefaultCollectionOrThrow() {
```

Comment:
I'm worried about this call.  It unconditionally cached a collection which could be later deleted.

Reply:
As default collection cannot be recreated, I think it's OK to reused the deleted one. I expected that LiteCore would handle this by throwing NotOpen.

Reply:
From the discussion:
1. We could simplify this by having only one getDefaultCollection(mustExist), and we could cache the default collection in when the database is created.
2. We could use c4db->hasCollection(spec) to check whether the collection exists or not in order to know if the collection's cache should be invalidated or not. When doing this, we could also refactor the code to getFromCache, check hashCollection() then get the c4Collection() if the cached one doesn't exist or is invalidated.